src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java

```diff
@@ -0,0 +1,505 @@
+                stageFiles.put(file, entry);
+            }
+            return this;
+        }
+
+        void updateFileStatus(Map<String, InternalUtils.IngestedFileStatus> report) {
+            report.forEach((fileName, status) -> stageFiles.computeIfPresent(fileName, (key, entry) -> {
+                entry.status = status;
+                return entry;
+            }));
+        }
+
+        boolean hasDirtyFiles() {
+            return !dirtyFiles.isEmpty();
+        }
+
+        boolean hasStageFiles() {
+            return !stageFiles.isEmpty();
+        }
+
+        Stream<String> query(Predicate<Map.Entry<String, IngestEntry>> filter) {
+            return stageFiles
+                    .entrySet()
+                    .stream()
+                    .filter(filter)
+                    .map(Map.Entry::getKey);
+        }
+
+        void stopTrackingFiles(List<String> files) {
+            files.forEach(file -> stageFiles.computeIfPresent(file, (key, entry) -> {
+                entry.keepTracking = false;
+                return entry;
+            }));
+        }
+    }
+
+    @VisibleForTesting
+    static class IngestEntry {
+        private InternalUtils.IngestedFileStatus status;
+        private final long timestamp;
+        private boolean keepTracking = true;
+
+        IngestEntry(InternalUtils.IngestedFileStatus status, long timestamp) {
+            this.status = status;
+            this.timestamp = timestamp;
+        }
+    }
+
+    static class ProgressRegisterImpl implements ProgressRegister {
+        private final LinkedBlockingQueue<String> files = new LinkedBlockingQueue<>();
+        private final AtomicLong offset = new AtomicLong(Long.MAX_VALUE);
+        private final SnowflakeTelemetryPipeCreation pipeCreation;
+        private final SnowflakeTelemetryPipeStatus pipeTelemetry;
+        private final  SnowflakeTelemetryService telemetryService;
+        private StageFilesProcessor owner = null;
+
+        ProgressRegisterImpl(SnowflakeTelemetryPipeCreation pipeCreation, SnowflakeTelemetryPipeStatus pipeTelemetry, SnowflakeTelemetryService telemetryService) {
+            this.pipeCreation = pipeCreation;
+            this.pipeTelemetry = pipeTelemetry;
+            this.telemetryService = telemetryService;
+        }
+
+        @Override
+        public void registerNewStageFile(String fileName) {
+            files.add(fileName);
+        }
+
+        @Override
+        public void newOffset(long offset) {
+            this.offset.set(offset);
+        }
+
+        @Override
+        public SnowflakeTelemetryPipeCreation getPipeCreation() {
+            return pipeCreation;
+        }
+
+        @Override
+        public SnowflakeTelemetryPipeStatus getPipeTelemetry() {
+            return pipeTelemetry;
+        }
+
+        @Override
+        public  SnowflakeTelemetryService getTelemetry() {
+            return telemetryService;
+        }
+
+        @Override
+        public void close() {
```

Comment:
Can we call `close` on `StageFilesProcessor` directly in `SnowflakeSinkServiceV1` and avoid having this awkward reference to an `owner` of this `ProgressRegisterImpl`?